emu_sv/sv_backend_impl.py

```diff
@@ -0,0 +1,248 @@
+import time
+import typing
+
+from pulser import Sequence
+import torch
+from resource import RUSAGE_SELF, getrusage
+
+from pulser.backend import Results, Observable, State, EmulationConfig
+from emu_base import PulserData
+
+from emu_sv.state_vector import StateVector
+from emu_sv.density_matrix_state import DensityMatrix
+from emu_sv.sv_config import SVConfig
+from emu_sv.time_evolution import EvolveStateVector
+from emu_sv.time_evolution import do_noisy_time_step
+
+
+class Statistics(Observable):
+    def __init__(
+        self,
+        evaluation_times: typing.Sequence[float] | None,
+        data: list[float],
+        timestep_count: int,
+    ):
+        super().__init__(evaluation_times=evaluation_times)
+        self.data = data
+        self.timestep_count = timestep_count
+
+    @property
+    def _base_tag(self) -> str:
+        return "statistics"
+
+    def apply(
+        self,
+        *,
+        config: EmulationConfig,
+        state: State,
+        **kwargs: typing.Any,
+    ) -> dict:
+        """Calculates the observable to store in the Results."""
+        assert isinstance(state, StateVector | DensityMatrix)
+        assert isinstance(config, SVConfig)
+        duration = self.data[-1]
+        if isinstance(state, StateVector) and state.vector.is_cuda:
+            max_mem_per_device = (
+                torch.cuda.max_memory_allocated(device) * 1e-6
+                for device in range(torch.cuda.device_count())
+            )
+            max_mem = max(max_mem_per_device)
+        elif isinstance(state, DensityMatrix) and state.matrix.is_cuda:
+            max_mem_per_device = (
+                torch.cuda.max_memory_allocated(device) * 1e-6
+                for device in range(torch.cuda.device_count())
+            )
+            max_mem = max(max_mem_per_device)
+        else:
+            max_mem = getrusage(RUSAGE_SELF).ru_maxrss * 1e-3
+
+        config.logger.info(
+            f"step = {len(self.data)}/{self.timestep_count}, "
+            + f"RSS = {max_mem:.3f} MB, "
+            + f"Δt = {duration:.3f} s"
+        )
+
+        return {
+            "RSS": max_mem,
+            "duration": duration,
+        }
+
+
+_TIME_CONVERSION_COEFF = 0.001  # Omega and delta are given in rad/μs, dt in ns
+
+
+class SVBackendImpl:
+
+    results: Results
+
+    def __init__(self, config: SVConfig, pulser_data: PulserData):
```

Comment:
I think we can share the code better between `SVBackendImpl` and `NoisySVBackendImpl`. For example, the lines up to here are identical in both `init` methods, so is the initialization of `self.results` and `self.statistics` and everything in the for loop, except for the function that does the time-step.

The parts that are not shared, are how the initial state is created, and what stepper is used. You could get those from methods `set_initial_state` and `get_stepper` which could have different implementations in both classes. Then you can share all the other code by using these functions.

Additionally, I think it's nicer if the for loop is in a `run` method.